### PR TITLE
Structured datastore query endpoint

### LINKED ISSUE #3073: Provide schema with machine and human-readable names in SQL endpoint response
Provide machine name and human readable name at a single endpoint

Show full schema in datastore import endpoint

with human readable names: https://demo.getdkan.org/api/1/datastore/sql/?query=%5BSELECT%20*%20FROM%20a5db2af0-3c43-51a9-908d-ee071030392c%5D%5BLIMIT%2020%20OFFSET%200%5D; 

with machine readable names: https://demo.getdkan.org/api/1/datastore/sql/?query=%5BSELECT%20*%20FROM%20a5db2af0-3c43-51a9-908d-ee071030392c%5D%5BLIMIT%2020%20OFFSET%200%5D;&show-db-columns

### LINKED ISSUE #3258: Handle CORS issues through standard Drupal config
The SQL endpoint brings in its own response config to allow CORS exceptions, but this isn't a sustainable strategy as we build out other APIs that we need to expose to decoupled applications. We should roll this back and allow the SQL and other endpoints to pick up Drupal's CORS settings, and allow people to configure request origin headers per-site and per-environment.

### PR DESCRIPTION
**Pull request by dafeder**
Introduces a new API endpoint for querying datastore tables using a JSON syntax. This need arose from a need for more functionality than the current SQL endpoint provides, including:

1. Support for mathematical expressions
2. Support for OR conditions and condition groups
3. Support for table joins

The decision to implement these first in a JSON endpoint rather than expanding the SQL endpoint came from these considerations:

1. SQL statements do not follow a set schema and are challenging to parse, don't want that to be the bottleneck
2. Building JSON objects is more natural/intuitive for frontend developers than building out SQL strings
3. We are working toward a new type of stored metadata for "filtered views"/"stored queries" and wanted to start building out the foundations of that. 

This PR also brings in our new library for working with validate-able JSON data structures in DKAN, GetDKAN/RootedJsonData. We are hoping this finds its way into the rest of the DKAN codebase and provides a standardized way to pass JSON data around the system and ensure that it is always valid. 

## QA Steps

1. Visit api/1/metastore/schemas/dataset/items?show-reference-ids and select the UUID of a particular distribution.
2. Try the following HTTP request, substituting your distribution ID for the one in this example:

```http
POST /api/1/datastore/query HTTP/1.1
Host: datastorebool.localtest.me
Content-Type: text/plain

{
    "resources": [
        {
            "id": "80503a94-77f4-560c-89bf-6f6984c9a137",
            "alias": "t"
        }
    ]
}
```

If this works, try different types of queries using the tests in this PR and the query.json schema for guidance. 

## Known issues

1. There are now two types of query objects being handled by the datastore - `DatastoreQuery` and the core DKAN `Query`. This is pretty redundant and adds a lot of complexity to translate from one to another. We should try to collapse or simplify this.
2. This PR removes a header that was previously present in DKAN REST API reponses allowing any origin, effectively bypassing CORS. DKAN endpoints will now simply set origin headers consistently with the rest of your Drupal site. For more information about handling CORS configuration in Drupal read [this thread](https://www.drupal.org/node/2715637).

## Merge steps

1. Merge GetDKAN/RootedJsonData#9
2. New release of RootedJsonData
2. Set https://github.com/GetDKAN/dkan/blob/query-joins-v1/composer.json#L27 to pull in new release
3. Merge

Resolves #3073, resolves #3258

### CODE REVIEW

**Comment by dafeder** [comment]
Having trouble understanding why the functional dataset test broke. I don't see any difference in relevant code between the two branches, but there is no `$%Ref:downloadURL` available on my branch. Here is a var_dump of the dataset from that point in the test:
```
object(stdClass)#5272 (10) {
  ["title"]=>
  string(7) "Test #1"
  ["description"]=>
  string(3) "Yep"
  ["identifier"]=>
  string(3) "123"
  ["accessLevel"]=>
  string(6) "public"
  ["modified"]=>
  string(10) "06-04-2020"
  ["keyword"]=>
  array(1) {
    [0]=>
    string(5) "hello"
  }
  ["distribution"]=>
  array(1) {
    [0]=>
    object(stdClass)#5268 (3) {
      ["title"]=>
      string(4) "blah"
      ["downloadURL"]=>
      string(91) "https://dkan-default-content-files.s3.amazonaws.com/phpunit/district_centerpoints_small.csv"
      ["mediaType"]=>
      string(8) "text/csv"
    }
  }
  ["%Ref:keyword"]=>
  array(1) {
    [0]=>
    object(stdClass)#5266 (2) {
      ["identifier"]=>
      string(36) "381f9022-e6a9-5d4d-8b54-b7771335505b"
      ["data"]=>
      string(5) "hello"
    }
  }
  ["%Ref:distribution"]=>
  array(1) {
    [0]=>
    object(stdClass)#5263 (2) {
      ["identifier"]=>
      string(36) "f9dbecd4-72e7-5fa5-91cb-a896a95179d2"
      ["data"]=>
      object(stdClass)#5261 (3) {
        ["title"]=>
        string(4) "blah"
        ["downloadURL"]=>
        string(91) "https://dkan-default-content-files.s3.amazonaws.com/phpunit/district_centerpoints_small.csv"
        ["mediaType"]=>
        string(8) "text/csv"
      }
    }
  }
  ["%modified"]=>
  string(10) "2020-11-08"
}
```

File: modules/common/src/JsonResponseTrait.php [hunk 1]
```diff
@@ -13,7 +13,7 @@ trait JsonResponseTrait {
    * Private.
    */
   private function getResponse($message, int $code = 200): JsonResponse {
-    return new JsonResponse($message, $code, ["Access-Control-Allow-Origin" => "*"]);
```
Comment (dafeder) [inline modules/common/src/JsonResponseTrait.php:16]: Document this change

File: modules/datastore/src/Service.php [hunk 2]
```diff
@@ -197,17 +227,165 @@ public function summary($identifier) {
...
+   *
+   * @return array
+   *   Array of result objects.
+   */
+  private function runResultsQuery(DatastoreQuery $datastoreQuery) {
+    $storageMap = $this->getQueryStorageMap($datastoreQuery);
+    $query = $this->populateQuery($datastoreQuery);
+    $primaryAlias = $datastoreQuery->{"$.resources[0].alias"};
+
+    if (!$primaryAlias) {
```
Comment (dafeder) [inline modules/datastore/src/Service.php:323]: Do this earlier

**Comment by grugnog** [comment]
Can one of the admins verify this patch?

**Comment by dafeder** [comment]
@fmizzell updated the sort format. Key commit is https://github.com/GetDKAN/dkan/pull/3248/commits/ae2cb7acd52eccd01321f79c12288a99d15c7211

**Pull request merged**